### ISSUE TITLE
Error bootstrapping toolkit on cluster with HelmRelease CRD v1

### ISSUE THREAD

**Issue by cdavisafc** [fluxcd/toolkit#110]
Trying to use the toolkit against a brand new K8s cluster

```
$ tk check --pre
► checking prerequisites
✔ kubectl 1.18.2 >=1.18.0
✔ Kubernetes 1.16.11 >=1.14.0
✔ prerequisites checks passed
```

But the bootstrap fails:

```
$ tk bootstrap github \
>  --owner=$GITHUB_USER \
>  --repository=gotk-example \
>  --path=gitops-runtime \
>  --personal
► connecting to github.com
✔ repository created
✔ repository cloned
✚ generating manifests
✔ components manifests pushed
► installing components in gitops-system namespace
namespace/gitops-system created
customresourcedefinition.apiextensions.k8s.io/alerts.notification.fluxcd.io created
customresourcedefinition.apiextensions.k8s.io/gitrepositories.source.fluxcd.io created
customresourcedefinition.apiextensions.k8s.io/helmcharts.source.fluxcd.io created
customresourcedefinition.apiextensions.k8s.io/helmrepositories.source.fluxcd.io created
customresourcedefinition.apiextensions.k8s.io/kustomizations.kustomize.fluxcd.io created
customresourcedefinition.apiextensions.k8s.io/providers.notification.fluxcd.io created
customresourcedefinition.apiextensions.k8s.io/receivers.notification.fluxcd.io created
role.rbac.authorization.k8s.io/crd-controller-gitops-system created
rolebinding.rbac.authorization.k8s.io/crd-controller-gitops-system created
clusterrolebinding.rbac.authorization.k8s.io/cluster-reconciler-gitops-system created
service/notification-controller created
service/source-controller created
service/webhook-receiver created
deployment.apps/helm-controller created
deployment.apps/kustomize-controller created
deployment.apps/notification-controller created
deployment.apps/source-controller created
networkpolicy.networking.k8s.io/deny-ingress created
The CustomResourceDefinition "helmreleases.helm.fluxcd.io" is invalid: status.storedVersions[0]: Invalid value: "v1": must appear in spec.versions
✗ install failed
```

**Comment by cdavisafc** [comment]
I just realized that this is probably because I am installing on a fresh WKP cluster - it already has a helmrelease CRD installed. Version conflict it seems.

**Comment by stefanprodan** [comment]
> Version conflict it seems.

Yes it's not possible to use helm-controller on WKP as helm-controller and helm-operator can't run on the same cluster. Even if you delete the helm-operator deployment and the HelmRelease v1 CRD, WKP's Flux will reapply them. The HelmRelease v2 is not backwards compatible with v1, so upgrading in-place is not an option, we need to write a v1-to-v2 convertor in `tk` cc @hiddeco

**Comment by paulcarlton-ww** [comment]
Is there a way to allow helm-controller and helm-operator to co-exist, maybe use a new crd name for helm-controller and have a migration tool that coverts existing helmrelease custom resources to the new format/name. 

My client is using helm-operator widely and wants to use GitOps toolkit source controller and kustomize-controller but is not yet convinced about replacing helm-operator with helm-controller. 

I suspect others have invested heavily in helm-operator and requiring an all or nothing migration feels like a risky approach.

At this stage I guess I could use tk install --export and remove helm-controller objects so I can deploy the bits of GitOps toolkit they want right now?

**Comment by hiddeco** [comment]
@paulcarlton-ww we will definitely investigate options at a later time but at the moment I am waiting for the various API design questions to be answered before committing to a migration tool.

**Comment by stefanprodan** [comment]
@paulcarlton-ww expect for source-controller all the others are optional components, if you want to bootstrap or install without helm-controller you can use the `components` arg like so:

```
tk install/bootstrap --components=source-controller,kustomize-controller,notification-controller
```

**Comment by paulcarlton-ww** [comment]
Great, would it be possible to add --components option to install too?

**Comment by stefanprodan** [comment]
@paulcarlton-ww check out the help command:

```
tk install -h
The install command deploys the toolkit components in the specified namespace.
If a previous version is installed, then an in-place upgrade will be performed.

Usage:
  tk install [flags]

Examples:
  # Install the latest version in the gitops-systems namespace
  tk install --version=latest --namespace=gitops-systems

  # Dry-run install for a specific version and a series of components
  tk install --dry-run --version=v0.0.7 --components="source-controller,kustomize-controller"

  # Dry-run install with manifests preview 
  tk install --dry-run --verbose

  # Write install manifests to file 
  tk install --export > gitops-system.yaml


Flags:
      --components strings   list of components, accepts comma-separated values (default [source-controller,kustomize-controller,helm-controller,notification-controller])
      --dry-run              only print the object that would be applied
      --export               write the install manifests to stdout and exit
  -h, --help                 help for install
      --manifests string     path to the manifest directory, dev only
  -v, --version string       toolkit version (default "latest")

```

**Comment by seaneagan** [comment]
I think there are two use cases:

1. Automated migration of HelmRelease v1 to v2.
2. Being able to consume both HelmRelease v1 and v2 in same cluster, during a migration period.

I would say 1. is a nice to have, but 2. is vital if you are consuming many HelmRelease documents, potentially from distributed sources, and cannot migrate them all at once.

**Comment by hiddeco** [comment]
They are still closely related.

The error as described in the issue is about `storedVersions`, once we have settled on a stable API we should declare `v2` as the `storedVersion` (as described in https://book.kubebuilder.io/multiversion-tutorial/api-changes.html) and write a converter for `v1` -> `v2` as described in https://book.kubebuilder.io/multiversion-tutorial/conversion.html.

**Comment by seaneagan** [comment]
I see, so there could be two approaches:

1. A migration tool which does a one-time conversion of source (e.g. stored in git) HelmReleases. And then using the strategy mentioned by @paulcarlton-ww of running the helm-operator (consuming v1) and helm-controller (consuming v2) in the same cluster.
2. A single controller which can convert v1 to v2 on the fly and thus consume both as mentioned by @hiddeco

They are not really in conflict with one another, but no need to double the effort.

**Comment by hiddeco** [comment]
> They are not really in conflict with one another, but no need to double the effort.

If we still want to offer both approaches, I think the converter logic of option 2 can be re-used in the utility made available for option. Which would greatly reduce the amount of work required to be able to offer both options to users.

**Comment by hiddeco** [comment]
Another reason for requiring option 1 is the fact that we will drop the support for values from external sources.

With an utility we would be able to fetch this external source and write the contents to a `ConfigMap` or `Secret` entry, while creating a `ValuesReference` in the `HelmRelease` to this resource. With the conversion happening automatically within the controller, this will be a harder task and likely impossible to pull off in a way that the contents are stored in Git.

**Comment by stefanprodan** [comment]
This has been fixed in the toolkit v0.0.13 release.

Now you can run both helm-operator and helm-controller in parallel and migrate from HelmRelease v1 to HelmRelease v2 gradually. The toolkit CRDs no longer conflict with Flux v1 as the group changed to `helm.toolkit.fluxcd.io`.
